[LLVM] [Clang] Support for Gentoo `*t64` triples (64-bit time_t ABIs)

mgorny opened the pull request:
Gentoo is planning to introduce a `*t64` suffix for triples that will be used by 32-bit platforms that use 64-bit `time_t`. Add support for parsing and accepting these triples, and while at it make clang automatically enable the necessary glibc feature macros when this suffix is used.

An open question is whether we can backport this to LLVM 19.x. After all, adding new triplets to Triple sounds like an ABI change — though I suppose we can minimize the risk of breaking something if we move new enum values to the very end.

mgorny committed:
[llvm] [Triple] Support *t64 environments for Gentoo

Add a support for the new `*t64` triples that Gentoo is planning to use
for its 32-bit platforms migrated to use 64-bit time_t.  Given
the supported 32-bit targets, the expected environments are `-gnut64`,
`-gnueabit64` and `-gnueabihft64`.

mgorny committed:
[clang] [Basic] Define time64 macros for *t64 triples

Set the `-D_FILE_OFFSET_BITS=64` and `-D_TIME_BITS=64` defaults when
clang is builting for one of the `*t64` triples that Gentoo is planning
to use for 32-bit platforms migrated to 64-bit time_t.

MaskRay commented on clang/test/Preprocessor/time64.c:
```diff
@@ -0,0 +1,12 @@
+// RUN: %clang_cc1 -E -dM -triple=i686-pc-linux-gnu < /dev/null | FileCheck -match-full-lines -check-prefix TIME32 %s
```
You can omit `< `

mgorny replied:
Yeah, I was wondering if it's in the `init.c` tests for a reason.

MaskRay commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -610,7 +610,8 @@ static llvm::Triple computeTargetTriple(const Driver &D,
         A->getOption().matches(options::OPT_maix64)) {
       AT = Target.get64BitArchVariant().getArch();
-      if (Target.getEnvironment() == llvm::Triple::GNUX32)
+      if (Target.getEnvironment() == llvm::Triple::GNUX32 ||
```
Optional: you could simplify the pattern  with `llvm::is_contained({...}, X)`

mgorny replied:
To be honest, I don't have the energy to rewrite it all manually, especially that these files don't seem to be clang-formatted consistently.

MaskRay commented on llvm/unittests/TargetParser/TripleTest.cpp:
```diff
@@ -1283,6 +1283,28 @@ TEST(TripleTest, ParsedIDs) {
+  EXPECT_EQ(Triple::UnknownVendor, T.getVendor());
+  EXPECT_EQ(Triple::Linux, T.getOS());
+  EXPECT_EQ(Triple::GNUEABIHFT64, T.getEnvironment());
+  EXPECT_TRUE(T.isTime64ABI());
```
needs a negative test for `isTime64ABI`

mgorny replied:
Added.

mgorny commented on clang/lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -177,6 +177,7 @@ createTargetCodeGenInfo(CodeGenModule &CGM) {
     else if (CodeGenOpts.FloatABI == "hard" ||
              (CodeGenOpts.FloatABI != "soft" &&
               (Triple.getEnvironment() == llvm::Triple::GNUEABIHF ||
```
I'm wondering if we should add a `isHardFloatABI()` or something like that to the `Triple` class.

MaskRay replied:
This helper sounds good.

will we have `muslt32`?

mgorny replied:
Are you asking if we're going to have musl suffixes on Gentoo? No, musl already switched to 64-bit `time_t`.

mgorny committed:
[llvm] [Triple] Streamline hardfloat test into isHardFloatABI()

Streamline checking for ARM hardfloat triplets into a `isHardFloatABI()`
method.  This keeps the list of environment matches in a single file,
rather than hardcoded into clang.
